filecache/__init__.py

```diff
@@ -269,65 +388,130 @@
-            if response.status_code == 200:
-                with open(local_path, 'wb') as f:
-                    f.write(response.content)
-                self._filecache._record_cached_file(filename, local_path)
-                return local_path
-            else:
-                raise FileNotFoundError(f'Failed to download file from: {url}')
+            return self._retrieve_from_web(filename, local_path)
 
         if self._source_type == 'gs':
-            blob_name = f'{self._prefix}{filename}'
-            blob = self._gs_bucket.blob(blob_name)
-            try:
-                blob.download_to_filename(str(local_path))
-            except (google.api_core.exceptions.BadRequest,  # bad bucket name
-                    google.resumable_media.common.InvalidResponse,  # bad bucket name
-                    google.cloud.exceptions.NotFound):  # bad filename
-                # The google API library will still create the file before noticing
-                # that it can't be downloaded, so we have to remove it here
-                local_path.unlink()
-                raise FileNotFoundError(
-                    f'Failed to download file from: gs://{self._gs_bucket_name}/'
-                    f'{blob_name}')
-            self._filecache._record_cached_file(filename, local_path)
-            return local_path
+            return self._retrieve_from_gs(filename, local_path)
 
         if self._source_type == 's3':
-            s3_key = f'{self._prefix}{filename}'
-            try:
-                self._s3_client.download_file(self._s3_bucket_name,
-                                              s3_key,
-                                              str(local_path))
-            except botocore.exceptions.ClientError:
-                raise FileNotFoundError(
-                    f'Failed to download file from: s3://{self._s3_bucket_name}/'
-                    f'{s3_key}')
-            self._filecache._record_cached_file(filename, local_path)
-            return local_path
+            return self._retrieve_from_s3(filename, local_path)
+
+        raise AssertionError(
+            f'Internal error unknown source type {self._source_type}')  # pragma: no cover
+
+    def _retrieve_from_web(self, filename, local_path):
+        url = f'{self._prefix}{filename}'
+        try:
+            response = requests.get(url, stream=True)
+            response.raise_for_status()
+        except requests.exceptions.RequestException as e:
+            raise FileNotFoundError(f'Failed to download file from: {url}') from e
+
+        temp_local_path = local_path.with_suffix(local_path.suffix + '.dltemp')
+        try:
+            with open(temp_local_path, 'wb') as f:
+                for chunk in response.iter_content(chunk_size=1024*1024):
+                    if chunk:
+                        f.write(chunk)
+            temp_local_path.rename(local_path)
+        except Exception:
```

Comment:
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/SETI/rms-filecache/pull/2#issuecomment-2377998362>#454d71</a><span> - 09/26/2024, 09:53 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> While the change from a bare except to except Exception is an improvement, it may still be too broad and catch exceptions that should be handled differently. A similar issue was also found in filecache/__init__.py (line 492).
</div>
<div id="fix">
<b>Fix:</b> Catch and handle specific exceptions that might occur during file operations, such as IOError or OSError, before the general Exception catch. This allows for more granular error handling and reporting.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -441,10 +441,14 @@
      temp_local_path = local_path.with_suffix(local_path.suffix + '.dltemp')
      try:
          with open(temp_local_path, 'wb') as f:
              for chunk in response.iter_content(chunk_size=1024*1024):
                  if chunk:
                      f.write(chunk)
          temp_local_path.rename(local_path)
 +    except (IOError, OSError) as e:
 +        temp_local_path.unlink(missing_ok=True)
 +        raise FileNotFoundError(f"Failed to write or rename file: {filename}") from e
      except Exception:
          temp_local_path.unlink(missing_ok=True)
          raise
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged